lib/IRGen/GenCall.cpp

```diff
@@ -3562,6 +3562,11 @@ llvm::CallBase *IRBuilder::CreateCallOrInvoke(
         builder.addByValAttr(func->getParamByValType(argIndex));
         attrs = attrs.addParamAttributes(func->getContext(), argIndex, builder);
       }
+      if (func->hasParamAttribute(argIndex, llvm::Attribute::InReg)) {
+        llvm::AttrBuilder builder(func->getContext());
+        builder.addAttribute(llvm::Attribute::InReg);
+        attrs = attrs.addParamAttributes(func->getContext(), argIndex, builder);
+      }
```

Comment:
Hmm.  I see that this is the established pattern we're using here, but it's clearly wrong and won't handle indirect calls correctly; we need to be setting these attributes based on the `CGFunctionInfo`.

Reply:
@rjmccall It's not obvious to me, can you elaborate on how to set these attributes based on the CGFunctionInfo?

Reply:
You want the attribute to already be present in the attributes returned by `fn.getAttributes()`, which ultimately means it should be added in `SignatureExpansion::expandExternalSignatureTypes` based on the same logic that Clang uses to do it.